src/pages/SigninPage/index.tsx

```diff
@@ -1,46 +1,98 @@
 import Jumbotron from '@/components/common/Jumbotron';
 import Logo from '@/components/common/Logo';
-import Shell from '@/components/shell/Shell';
 import { NETWORK_ERROR, UNEXPECTED_ERROR, loginError } from '@/constants/error';
 import Uri from '@/constants/uri';
+import useShellScript from '@/hooks/useShellScript';
 import { LoginRequestBody } from '@/types';
+import { PrintLineCommand } from '@/types/shell';
 
 import { footerStyle, headerStyle, layoutStyle } from './style';
 
 const SigninPage: FC = () => {
   const { push } = useHistory();
 
-  const requestSignin = async (body: LoginRequestBody) => {
+  const [
+    signinScript,
+    addPrintLineCommand,
+    addReadLineCommand,
+    createPrintLineCommand,
+  ] = useShellScript();
+
+  const checkUserName = async (val: string) => {
+    if (val.length) return { status: 'success', nextSequence: 2 } as const;
+
+    const errorLine = createPrintLineCommand(1, 'error', 'Please enter your ID');
+    return { status: 'error', ...errorLine } as const;
+  };
+
+  const checkPassword = async (val: string) => {
+    if (val.length) return { status: 'success', nextSequence: 3 } as const;
+
+    const errorLine = createPrintLineCommand(2, 'error', 'Please enter your Password');
+    return { status: 'error', ...errorLine } as const;
+  };
+
+  const checkSigninForm = async (
+    val: string,
+    body: LoginRequestBody,
+  ): Promise<
+    | {
+        status: 'success';
+        nextSequence: number | null;
+      }
+    | ({
+        status: 'error';
+      } & PrintLineCommand)
+  > => {
+    if (val !== 'y') {
+      const errorLine = createPrintLineCommand(1, 'error', `Access denied for user ${body.userId}`);
+      return { status: 'error', ...errorLine } as const;
+    }
+
     try {
       await authApi.login(body);
       push(Uri.service);
     } catch (error) {
-      if (!error.response) throw new Error(NETWORK_ERROR);
+      if (!error.response) {
+        const errorLine = createPrintLineCommand(1, 'error', NETWORK_ERROR);
+        return { status: 'error', ...errorLine } as const;
+      }
 
       const { status } = error.response;
-      if (loginError[status]) throw new Error(loginError[status]);
+      if (loginError[status]) {
+        const errorLine = createPrintLineCommand(1, 'error', loginError[status]);
+        return { status: 'error', ...errorLine } as const;
+      }
 
-      throw new Error(UNEXPECTED_ERROR);
+      const errorLine = createPrintLineCommand(1, 'error', UNEXPECTED_ERROR);
+      return { status: 'error', ...errorLine } as const;
     }
+
+    return { status: 'success', nextSequence: null };
   };
 
+  addPrintLineCommand(0, 1, '', 'telbby init v0.1.0');
+  addReadLineCommand(1, 'question','username', checkUserName, null, 'userId', 30);
+  addReadLineCommand(2, 'question', 'password', checkPassword, null, 'password', 35);
+  addReadLineCommand(3, 'question', 'Sign in? [y/n]', checkSigninForm, 'y', null, 1);
```

Comment:
개인적으로 네이밍을 바꾸는게 나아보입니다 !
처음보는 저의 입장에서는 PrintLine이 오히려 작성하는 shell 을 설정하는 것처럼 느껴지네요

그리고 리턴타입이 없는 함수가 저는 조금 별로인 것 같아요.. 
만약 외부로 부터 shellLine을 받는다면, 커스텀 훅을 통해 script에 대한 리턴을 주고, 이를 ShellLine의 props로 전달하면 좋을 것 같습니다.

```suggestion
  const firstScript = useReadLineCommand(0, 1, '', 'telbby init v0.1.0');
  const secondScript = useWriteLineCommand(1, 'question','username', checkUserName, null, 'userId', 30);
  const thirdScript = WriteLineCommand(2, 'question', 'password', checkPassword, null, 'password', 35);
  const forthScript = useWriteLineCommand(3, 'question', 'Sign in? [y/n]', checkSigninForm, 'y', null, 1);
```

이후 이러한 Script를 컴포넌트 리턴하는 곳에서 아래와 같은 방식으로 사용하는 건 어떤가요 ???

```tsx
<Shell>
  <ShellLine script={firstScript}
</Shell>
```